### ISSUE TITLE
Implement SmvOutputBase

### ISSUE THREAD

**Issue by ninjapapa** [TresAmigosSD/SMV#1202]
`SmvOutputBase` is the base class of all future output modules. We will move away from the `publish` and `publish-hive` approach. Instead we will create specific output modules (derived from `SmvOutputBase`, with the only task of those modules are write the DF out to somewhere.

**Comment by laneb** [comment]
The difference between publishing to hive and publishing to csv ought to be limited to data store itself. The semantics of publishing data ought ought to be the same regardless of data store. It should be easy for the output to be read back with SMV. E.g., if the user publishes a module to Hive, they shouldn't have to create a `HiveInputTable` to read the table back in. They ought to be able to link to the published module and read the published data automatically.

**Comment by ninjapapa** [comment]
The "it should be easy to read back" requirement does not fit into the original idea of SmvOutputBase. @laneb could you explain it a little more? Is it for the current multi-stage + publish use cases?

**Comment by ninjapapa** [comment]
Sounds like there are 2 different "output" requirements:
* For delivery: reporting, other data applications (not even Spark), etc.
* For persisting in a form can be investigated for development, QA and coordination between teams, other than using SMV 

May need to address those 2 separately instead of try to accomplish by the same thing.

**Comment by laneb** [comment]
Can you elaborate on the purpose of `SmvOutputBase`? I thought that the purpose was to distinguish the data store that a module will write to when published and means of writing to that data store. Is the idea that `SmvOutputBase` will actually have a different behavior when _persisting_ output?

**Comment by laneb** [comment]
@AliTajeldin Okay, I'm with you there. Not following how the hash got involved, though. It is only relevant when dealing with persisted output, not published. I suspect there is a conceptual part of this plan I'm missing. What if I want to link to an `SmvHiveOutput`? Is the idea to get rid of the current publish/link semantics entirely?

**Comment by AliTajeldin** [comment]
@laneb Correct.  No more link stuff.  All gone.  Was too confusing for users.  They just published to hive and used the output table as the input in the next "stage/project"
